File: csharp-workshop/Part2.LightMocks.Tests/Service/PaymentServiceTest.cs
using NSubstitute;
using Part2.LightMocks.Domain;
using Part2.LightMocks.Gateway;
using Part2.LightMocks.Service;
using Xunit;

namespace Part2.LightMocks.Tests.Service;

public class PaymentServiceTest
{
    [Fact]
    public void P2_ProcessPayment_ReturnsSuccess_WhenGatewaySucceeds()
    {
        var paymentGateway = Substitute.For<IPaymentGateway>();
        var paymentService = new GatewayPaymentService(paymentGateway);

        paymentGateway.Process(Arg.Any<PaymentRequest>())
            .Returns(new PaymentResult("", PaymentStatus.Success, "TXN-789"));

        var result = paymentService.ProcessPayment("ORDER-123", "CUST-456", 100.0);

        Assert.Equal(PaymentStatus.Success, result.Status);
        Assert.Equal("ORDER-123", result.OrderId);
        Assert.Equal("TXN-789", result.TransactionId);
        paymentGateway.Received(1).Process(Arg.Is<PaymentRequest>(request =>
            request.OrderId == "ORDER-123"
            && request.CustomerId == "CUST-456"
            && request.Amount == 100.0
        ));
    }

    [Fact]
    public void P2_ProcessPayment_ReturnsFailure_ForZeroAmount_WithoutCallingGateway()
    {
        var paymentGateway = Substitute.For<IPaymentGateway>();
        var paymentService = new GatewayPaymentService(paymentGateway);

        paymentGateway.Process(Arg.Any<PaymentRequest>())
            .Returns(new PaymentResult("", PaymentStatus.Success, "TXN-789"));

        var result = paymentService.ProcessPayment("ORDER-123", "CUST-456", 0.0);

        Assert.Equal(PaymentStatus.Failed, result.Status);
        Assert.Equal("ORDER-123", result.OrderId);
        paymentGateway.DidNotReceiveWithAnyArgs();
    }

    [Fact]
    public void P2_ProcessPayment_ReturnsFailure_WhenGatewayFails()
    {
        var paymentGateway = Substitute.For<IPaymentGateway>();
        var paymentService = new GatewayPaymentService(paymentGateway);

        paymentGateway.Process(Arg.Any<PaymentRequest>())
            .Returns(new PaymentResult("", PaymentStatus.Failed, null));

        var result = paymentService.ProcessPayment("ORDER-123", "CUST-456", 100.0);

        Assert.Equal(PaymentStatus.Failed, result.Status);
        Assert.Null(result.TransactionId);
    }

    [Fact]
    public void P2_RefundPayment_DelegatesToGateway()
    {
        var paymentGateway = Substitute.For<IPaymentGateway>();
        var paymentService = new GatewayPaymentService(paymentGateway);

        paymentGateway.Refund("TXN-123").Returns(true);

        Assert.True(paymentService.RefundPayment("TXN-123"));
        paymentGateway.Received(1).Refund("TXN-123");
    }
}

File: typescript-workshop/part-2-light-mocks/tests/service/PaymentServiceTest.ts
import { describe, it, expect, vi } from 'vitest';
import { PaymentRequest } from '../../src/domain/PaymentRequest';
import { PaymentResult } from '../../src/domain/PaymentResult';
import { PaymentStatus } from '../../src/domain/PaymentStatus';
import type { PaymentGateway } from '../../src/gateway/PaymentGateway';
import { GatewayPaymentService } from '../../src/service/GatewayPaymentService';

describe('PaymentService', () => {
  it('p2_processPayment_returnsSuccess_whenGatewaySucceeds', () => {
    const paymentGateway = { process: vi.fn(), refund: vi.fn() } as unknown as PaymentGateway;
    const paymentService = new GatewayPaymentService(paymentGateway);

    (paymentGateway as any).process.mockReturnValue(new PaymentResult('', PaymentStatus.Success, 'TXN-789'));

    const result = paymentService.processPayment('ORDER-123', 'CUST-456', 100.0);

    expect(result.status).toBe(PaymentStatus.Success);
    expect(result.orderId).toBe('ORDER-123');
    expect(result.transactionId).toBe('TXN-789');
    expect((paymentGateway as any).process).toHaveBeenCalledWith(
      expect.objectContaining({
        orderId: 'ORDER-123',
        customerId: 'CUST-456',
        amount: 100.0,
      }),
    );
  });

  it('p2_processPayment_returnsFailure_forZeroAmount_withoutCallingGateway', () => {
    const paymentGateway = { process: vi.fn(), refund: vi.fn() } as unknown as PaymentGateway;
    const paymentService = new GatewayPaymentService(paymentGateway);

    (paymentGateway as any).process.mockReturnValue(new PaymentResult('', PaymentStatus.Success, 'TXN-789'));

    const result = paymentService.processPayment('ORDER-123', 'CUST-456', 0.0);

    expect(result.status).toBe(PaymentStatus.Failed);
    expect(result.orderId).toBe('ORDER-123');
    expect((paymentGateway as any).process).not.toHaveBeenCalled();
  });

  it('p2_processPayment_returnsFailure_whenGatewayFails', () => {
    const paymentGateway = { process: vi.fn(), refund: vi.fn() } as unknown as PaymentGateway;
    const paymentService = new GatewayPaymentService(paymentGateway);

    (paymentGateway as any).process.mockReturnValue(new PaymentResult('', PaymentStatus.Failed, null));

    const result = paymentService.processPayment('ORDER-123', 'CUST-456', 100.0);

    expect(result.status).toBe(PaymentStatus.Failed);
    expect(result.transactionId).toBeNull();
  });

  it('p2_refundPayment_delegatesToGateway', () => {
    const paymentGateway = { process: vi.fn(), refund: vi.fn() } as unknown as PaymentGateway;
    const paymentService = new GatewayPaymentService(paymentGateway);

    (paymentGateway as any).refund.mockReturnValue(true);

    expect(paymentService.refundPayment('TXN-123')).toBe(true);
    expect((paymentGateway as any).refund).toHaveBeenCalledWith('TXN-123');
  });
});
